### PR TITLE
Wiki issue 155

### LINKED ISSUE #155: There is no development guide for how to implement offline capabilities
**Is your feature request related to a problem? Please describe.**
As a developer I have no clue how to implement the application offline.

**Describe the solution you'd like**
A clear and concise description of how to implement the application offline. The method chosen should be reviewed and discussed with the other development teams.

### PR DESCRIPTION
**Pull request by aerlingsson**

### CODE REVIEW

**Review by melanieselman — DISMISSED** [review]

File: docs/development/Offline_mode.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+1. Citizen 1 logs in on their device and downloads their weekplans from the server.
+2. Guardian logs in on another device and changes Citizen 1's weekplan for a week.
+3. Citizen 1 looks at this weekplan on their device, but this is not the updated version since there is a version in the cache.
+
+Possible solutions is to e.g. use a timestamp to check if there is changes in the online version. This timestamp would be downloaded and checked whenever a weekplan is opened on a device with an internet connection. Then this timestamp is compared with the local version.   
+Another solution would be to automatically check for changes every e.g. 30 seconds to avoid having to reopen weekplans to update them. A guardian could also have a "Refresh" button for the citizens, that would download the new changes.    
+Time stamps could also solve the update conflicts since it is possible to compare to versions and save the newest.   
+   
+*Synchronizing offline changes scenario:*
+1. Citizens table is offline.
```
Comment (giano574) [inline docs/development/Offline_mode.md]:
```suggestion
1. Citizen's tablet is offline.   
```

File: docs/development/Offline_mode.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+2. Guardian logs in on another device and changes Citizen 1's weekplan for a week.
+3. Citizen 1 looks at this weekplan on their device, but this is not the updated version since there is a version in the cache.
+
+Possible solutions is to e.g. use a timestamp to check if there is changes in the online version. This timestamp would be downloaded and checked whenever a weekplan is opened on a device with an internet connection. Then this timestamp is compared with the local version.   
+Another solution would be to automatically check for changes every e.g. 30 seconds to avoid having to reopen weekplans to update them. A guardian could also have a "Refresh" button for the citizens, that would download the new changes.    
+Time stamps could also solve the update conflicts since it is possible to compare to versions and save the newest.   
+   
+*Synchronizing offline changes scenario:*
+1. Citizens table is offline.   
+2. Guardians logs in and changes the citizens settings.
```
Comment (giano574) [inline docs/development/Offline_mode.md]:
```suggestion
2. A guardian logs in on the same (offline) tablet and changes the citizen's settings.
```

File: docs/development/Offline_mode.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+3. The citizen logs in and can see the local updates.
+4. The citizens device gets internet connection and now the changes has to be synced with the database. But now the guardian is not logged in anymore and the citizen do not have the permission to update their changes through the web-api.
```
Comment (giano574) [inline docs/development/Offline_mode.md]:
```suggestion
4. The citizen's device gets internet connection and now the changes has to be synced with the database. But the guardian is not logged in anymore and the citizen does not have permission to update their changes through the web-api.

```

File: docs/development/Offline_mode.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+   
+Possible solution would be to give a citizen more permissions in the wep-api to make changes
```
Comment (giano574) [inline docs/development/Offline_mode.md]:
```suggestion
A possible solution would be to give a citizen permissions in the wep-api to make changes.

```

File: docs/development/Offline_mode.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+**Flutter resources:**    
+https://github.com/tekartik/sqflite/blob/master/sqflite/README.md  
+https://pub.dev/packages/connectivity   
+https://flutter.dev/docs/cookbook/images/cached-images   
+https://pub.dev/packages/cached_network_image
+
+## Our solution suggestion
+We propose what is called an offline first approach because if Giraf wants to be a widely used app on different app markets, it is expected to be highly functional without internet connection. Offline first also consumes less battery and data - where especially long battery time is important for institutions and families going on trips.
+
+![image](https://cdn.discordapp.com/attachments/674940529124966402/691622267402190858/unknown.png)
```
Comment (giano574) [inline docs/development/Offline_mode.md]: The image is a little misleading; When the memory cache and local database can't respond to a query, and the REST API returns an answer, the path back to the bloc from the model adapter (in the lower right corner) is unclear. This was mentioned in the feedback by group 15.

**Review by giano574 — CHANGES_REQUESTED** [review]
Generally fine

**Review by giano574 — APPROVED** [review]

**Review by melanieselman — APPROVED** [review]

**Pull request merged**